### ISSUE TITLE
how can i use this model in imu/gnss navigation?

### ISSUE THREAD

**Issue by JiangWeiHn** [UTS-RI/ugpm#7]
I want use this ugpm to replace the normal preintegration in the imu/gnss Graph optimization system.  I want know every imu time the body’s position, velocity and attitude。so i set the start_t is the first imu time,and infer_t is the second imu time .and the preint_opt is :
    opt_.type = ugpm::PreintType::UGPM;
    opt_.min_freq = 500;
    opt_.state_freq = 400;
    opt_.quantum = -1.0;
my imu frequency is 400Hz.

and this is the code to create the preintegration class:
   imu_for_preint_.acc.push_back(imu_data_.acc.at(imu_id-1));
    imu_for_preint_.acc.push_back(imu_data_.acc.at(imu_id));
    imu_for_preint_.gyr.push_back(imu_data_.gyr.at(imu_id-1));
    imu_for_preint_.gyr.push_back(imu_data_.gyr.at(imu_id));
    double start_t = imu_for_preint_.acc[0].t;
    double infer_t = imu_for_preint_.acc[1].t;
 ugpm::ImuPreintegration preint(imu_for_preint_,start_t,infer_t,opt_,prior_bias);

and I got:
start_t:350110.3100683689 infer_t:350110.3121600151,this is the create the preint class ,the  start_t value,and the infer_t value.
at  the IterativeIntegrator,in the function velPosPreintLPMPartial,i change the code  as follow:
              
while(acc_time_[data_ptr+1] < start_t_){
                    data_ptr++;
                    if(data_ptr == (nb_acc_-1)){
                        std::cout<<std::fixed<<std::setprecision(10)<< "start_t::"<<start_t_<<std::endl;
                        std::cout<< std::fixed<<std::setprecision(10)<<"acc_time_[data_ptr+1]::"<<acc_time_[data_ptr]<<std::endl;
                        throw std::range_error("LPM Partial: the start_time is not in the data domain");
                    }
                }

and got:
 
start_t::350110.2933351994
acc_time_[data_ptr+1]::350110.2921600151
terminate called after throwing an instance of 'std::range_error'
  what():  LPM Partial: the start_time is not in the data domain
[ob_gins_node-2] process has died [pid 4549, exit code -6, cmd /home/slamer/workspace/ugpm_ob_gins/devel/lib/ob_gins/ob_gins_ros __name:=ob_gins_node __log:=/home/slamer/.ros/log/8fdefe88-8c37-11ef-8357-dfe18a224d4d/ob_gins_node-2.log].
I noticed that there are two start_t_ times，one is in the ImuPreintegration,and other in the class  IterativeIntegrator, they have different value,why this happen? and how can i fixed this bug?   look forward to your reply,thanks a lot!!

**Comment by clegenti** [comment]
Did you try to provide more data to the imu data structure (like a bigger overlap before and after the start and end times)?

One thing that I don't fully get is why you try to generate preintegration between consecutive IMU measurements? The idea of preintegration is to combined IMU measurements together to create pseudo measurements.
What is the delta t between consecutive states of your imu/gnss factor graph?

**Comment by JiangWeiHn** [comment]
The program ran for less than 1 second before reporting this bug.
the gnss signal is 1Hz。If the navigation outputs navigation results in 1 second, it cannot meet the requirements for SLAM。so I referred to this project，https://github.com/i2Nav-WHU/OB_GINS.     And in this project,when gnss signal comes.It combined the marginalization,imu residual，gnss residual in the graph optimize。In other cases, when IMU data is available, navigation results are produced using normal preintegration.
I set the imu_for_preint_ length 4.And I set the kNumDtJacobianDelta = 0.0025 ,which corresponding to the frequency of my IMU.The program can run now，but the output of  position increment is very small,can't fuse with GNSS signal.I subtracted the gravitational acceleration from the output of the accelerometer，is that right?
